CheckDupWithinSilly.GCL.r

```diff
@@ -1,127 +1,281 @@
+    
+    get(paste(silly, ".gcl", sep=""), pos = 1)
+    
+    }) %>% purrr::set_names(sillyvec)
+  
+  #Start if NULL quantile
   if(is.null(quantile)){
-    require("reshape")
-    library("dplyr")
-    for(silly in sillyvec){
-      my.gcl=get(paste(silly,".gcl",sep=""),pos=1)
-      scores=my.gcl$scores[,loci,,drop=FALSE]
-      ID=as.character(dimnames(scores)[[1]])
-      n=length(ID)
-      if(n<2){
-        resultlist[[silly]]=list(report="No Duplicates",DupDist=NULL) 
-        next()
-      }
-      scores.df=data.frame(t(sapply(ID,function(id){sapply(loci,function(locus){ifelse(is.na(scores[id,locus,1]),NA,paste(scores[id,locus,1:ploidy[locus]],collapse="."))})})))  
-      sort.scores.df=sort_df(scores.df)
-      sortIDs=dimnames(sort.scores.df)[[1]]
-      duplication=t(sapply(1:(n-1),function(id){ 
-                    vec=sapply(1:nloci,function(locus){ 
-                          sort.scores.df[id,locus]==sort.scores.df[id+1,locus]
-                        });
-                    ifelse(sum(!is.na(vec)),sum(vec[!is.na(vec)])/sum(!is.na(vec)),0 )
-                    }))
-      dupIND=as.vector(duplication>minproportion)
+    
+    cl <- parallel::makePSOCKcluster(ncores)
+    
+    doParallel::registerDoParallel(cl, cores=ncores)  
+    
+    # multicore loop
+    
+    dupcheck <- foreach::foreach(silly = sillyvec, .packages = "tidyverse") %dopar% {
+      
+      new.gcl <- my.gcl[[silly]]
+        
+      IDs <- new.gcl$FK_FISH_ID
+      
+      n <- dim(new.gcl)[1]
+      
+      nloci <- length(loci)
+      
+      #Combine dose 1 and 2 into single columns for each locus separated by a period.
+      scores0 <- new.gcl[ ,c("FK_FISH_ID", scores_cols)] 
+    
+      if(n < 2){
+        
+        report <- tibble::tibble(silly = silly, ID1 = NA, ID2 = NA, Missing1 = NA, Missing2 = NA, proportion = NA) #If only one individual the results will be a tibble of NAs. Can't do next() using applys
+        
+      } else{
+        
+        scores1 <- sapply(IDs, function(id){
+          
+          ind <- scores0 %>% 
+            filter(FK_FISH_ID==id)
+          
+          sapply(loci,function(locus){
+            
+            ifelse(is.na(ind[ ,locus]), NA, paste(ind[ , locus], ind[ , paste0(locus,".1")], sep = "."))
+            
+            })
+          
+          }) %>% 
+          t() %>% 
+          data.frame()
+        
+        dimnames(scores1)[[1]] <- IDs
+        
+        sort.scores.df <- reshape::sort_df(scores1)
+        
+        sortIDs <- dimnames(sort.scores.df)[[1]]
+        
+     #Calculate the porpotion of duplicated loci for each silly
+      duplication <- t(sapply(1:(n-1), function(id){ 
+            
+        vec <- sapply(1:nloci, function(locus){ 
+          
+          sort.scores.df[id, locus]==sort.scores.df[id+1, locus]
+          
+        });
+        
+        ifelse(sum(!is.na(vec)), sum(vec[!is.na(vec)])/sum(!is.na(vec)), 0 )
+        
+      }))

```

Comment:
This is a bug, after sorting the entire data.frame, it only looks for duplicate fish on either side. If two duplicate fish have NA's at different loci, they may not sort next to each other, and thus may be missed by this lazy look.

Reply:
This bug has been in here from the beginning, from the looks of it.

Reply:
You can test this yourself if you run the function with `quantile = NULL` and with `quantile = 0.99` and you can end up with different numbers of duplicates.